Support for multiple forcefields

arepstein opened the pull request:
Improvements on forcefield specification for mixed systems and specific water models

## Summary

Include a summary of major changes in bullet points:

* Added capability to use base names of forcefields, e.g. "gaff" or "spce"
* Added partial charges to multiple forcefield system setup
* Converted all system and forcefield creation to OpenMM rather than OpenFF
* Added test for system with spce water and charges specified


## TODO (if any)

* Unsure of how units should be handled in specifying box size

## Checklist

Work-in-progress pull requests are encouraged, but please put [WIP]
in the pull request title.

Commit:
Support for multiple forcefields

Improvements on forcefield specification for mixed systems and specific water models

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -341,130 +365,247 @@ def assign_charges_to_mols(
+    basic_water_ffs = ["tip3p", "spce", "tip4p"]
+    basic_small_ffs = ["gaff", "sage"]
+
+    all_small_ffs = [
```
Could we get rid of this list and make list generation programmatic?

arepstein replied:
Should be good now

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -341,130 +365,247 @@ def assign_charges_to_mols(
+        partial_charge_method: Method for OpenFF partial charge assignment
+                                for small molecules without charges provided
+                                in partial_charges
+        partial_charge_scaling: Scaling for partial charges, either
```
this actually cannot be done uniformly as a float, otherwise this is right

arepstein replied:
fixed docstring

Commit:
pre-commit run

Implemented changes in comments, ran pre-commit

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -360,111 +360,182 @@ def parameterize_system(
+            partial_charges=partial_charges,
+            partial_charge_scaling=partial_charge_scaling,
+            partial_charge_method=partial_charge_method,
+        )
```
how about define `ff_name = force_field.lower()` up here and then use `ff_name` throughout?

arepstein replied:
Mostly implemented. Apart from large molecules that need capitals.

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -360,111 +360,182 @@ def parameterize_system(
-            if ff_name.lower() in small_ffs:
-                small_molecules[openff_mol] = ff_name.lower()
+            found_isomorphic = False
+            if len(partial_charges) > 0:
+                for mol, charges in partial_charges:
+                    inferred_mol = infer_openff_mol(mol)
+                    is_isomorphic, atom_map = get_atom_map(inferred_mol, openff_mol)
+                    if is_isomorphic:
+                        charged_mol = assign_charges_to_mols(
+                            smile_strings=[smile],
+                            partial_charges=[(mol, charges)],
+                            partial_charge_method=partial_charge_method,
+                            partial_charge_scaling=partial_charge_scaling,
+                        )[0]
+                        found_isomorphic = True
+                if found_isomorphic:
+                    mol_to_load = charged_mol
+                else:
+                    mol_to_load = openff_mol
```
So if you provide a list of smiles to `assign_charges_to_mol` it will return a list of `charged_mols`. Parameterized from the provided charges if it can find an isomorphism and parameterized from `partial_charge_method` if not. It seems like you could call `assign_charges_to_mols` on the whole list rather than loop through all the mols one-by-one here? You are guaranteed to end up with charges versions of all the smiles you provide.

orionarcher replied:
You could then create a dict of {smile: charged_mol} and iterate through that in the big loop

arepstein replied:
Implemented.

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -360,111 +360,182 @@ def parameterize_system(
+            partial_charge_scaling=partial_charge_scaling,
+            partial_charge_method=partial_charge_method,
+        )
+        if force_field.lower() in basic_small_ffs:
+            if force_field.lower() == "sage":
+                sage = SMIRNOFFTemplateGenerator(molecules=charged_off_mols, forcefield="openff-2.0.0")
+                forcefield_omm.registerTemplateGenerator(sage.generator)
+            elif force_field.lower() == "gaff":
+                gaff = GAFFTemplateGenerator(molecules=charged_off_mols, forcefield="gaff-2.11")
+                forcefield_omm.registerTemplateGenerator(gaff.generator)
+        else:
+            if "sage" in force_field.lower():
+                sage = SMIRNOFFTemplateGenerator(molecules=charged_off_mols, forcefield=force_field.lower())
+                forcefield_omm.registerTemplateGenerator(sage.generator)
+            elif "gaff" in force_field.lower():
+                gaff = GAFFTemplateGenerator(molecules=charged_off_mols, forcefield=force_field.lower())
+                forcefield_omm.registerTemplateGenerator(gaff.generator)
+
```
How about 
```python
smirnoff_ff_names = SMIRNOFFTemplateGenerator.INSTALLED_FORCEFIELDS
gaff_ff_names = GAFFTemplateGenerator.INSTALLED_FORCEFIELDS
if  ff_name == "sage" or ff_name in smirnoff_ff_names:
   ff_name = "openff-2.0.0" if ff_name == "sage" else ff_name
   ... do sage stuff
elif ff_name == "gaff" or ff_name in gaff_ff_name:
   ff_name = "gaff-2.11" if ff_name == "gaff" else ff_name
   ... do gaff stuff
else:
  ... # something about the forcefield not being implemented?
```

orionarcher replied:
I'd also recommend wrapping this in a function, maybe `assign_small_molecule_ff` so you can use it down below.

arepstein replied:
Do you think this should actually throw an Error rather than a Warning about the forcefield not being implemented? Or should we default to sage and just warn?

orionarcher replied:
I'd vote for an Error, someone might miss a warning in high-throughput. If a FF is not available, I would think that is something that needs to be addressed. What would your preference be?

arepstein replied:
Agreed, put in an error for now

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -360,111 +360,182 @@ def parameterize_system(
                 gaff = GAFFTemplateGenerator(molecules=mol, forcefield=ff)
                 forcefield_omm.registerTemplateGenerator(gaff.generator)
-            elif "smirnoff" in ff or "openff" in ff:
+            elif "smirnoff" in ff_name or "openff" in ff_name or "sage" in ff_name:
+                if ff_name.lower() == "sage":
+                    ff = "openff-2.0.0"
+                else:
+                    ff = ff_name
```
maybe a case for using a `assign_small_molecule_ff` function

arepstein replied:
Added `assign_small_molecule_ff`

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -360,111 +360,182 @@ def parameterize_system(
+                small_molecules[mol_to_load] = "sage"
+
+        # Determine category of forcefield for deciding which water model
+        large_ff_category = ""
```
I'd recommend using `None` rather than an empty string

arepstein replied:
Done

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -360,111 +360,182 @@ def parameterize_system(
+                if ff_name.lower() in all_small_ffs or ff_name.lower() in basic_small_ffs:
+                    small_molecules[mol_to_load] = ff_name.lower()
+                else:
+                    large_or_water[mol_to_load] = ff_name.lower()
```
I actually think that "large" isn't really a category here. My understanding is that "small molecule" force fields use "small" to distinguish them from coarse grained or protein force fields, since we aren't supporting those, I think that our molecules are either "small molecules" or water. I think that if we reorganized this a bit to treat water as one case and not-water as another, it might be a bit cleaner.

arepstein replied:
I'm using "large" to allow for Amber and CHARMM DNA and protein models.

orionarcher replied:
Got it, thanks for clarifying. In that case, ignore my comment. Hopefully someone will use this for large molecules lol.

On that note "large molecules" doesn't feel like the right term. What would be more appropriate? Biomolecules? Should we change it in the code?

arepstein replied:
Yeah I wish I knew the field better. Based on the OpenMM documentation, 

> "The OpenMM force fields above include pregenerated templates for biopolymers and solvents. If your system instead contain small molecules, it is often necessary to generate these parameters on the fly."

So may "biopolymers_and_solvents" might be best

orionarcher replied:
I like what you settled on here.

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -360,111 +360,182 @@ def parameterize_system(
-            forcefield_omm.loadFile(ff)
-        for mol, ff in small_molecules.items():
-            if "gaff" in ff:
+            if ff in basic_water_ffs:
+                # Ensure the water model matches the large molecule model
+                if large_ff_category:
+                    if large_ff_category == "amber":
+                        if ff == "spce":
+                            ff_to_load = "amber14/spce.xml"
+                        if ff == "tip3p":
+                            ff_to_load = "amber14/tip3p.xml"
+                        if ff == "tip4p":
+                            ff_to_load = "amber14/tip4pew.xml"
+                    elif large_ff_category == "charmm":
+                        if ff == "spce":
+                            ff_to_load = "charmm36/spce.xml"
+                        if ff == "tip3p":
+                            ff_to_load = "charmm36/water.xml"
+                        if ff == "tip4p":
+                            ff_to_load = "charmm36/tip4pew.xml"
+                    else:
+                        warnings.warn(f"Did you mean to use {ff} with the " f"{large_ff_category} force field?")
+                # If there isn't a large molecule forcefield required,
+                # assume amber14
+                else:
+                    if ff == "spce":
+                        ff_to_load = "amber14/spce.xml"
+                    if ff == "tip3p":
+                        ff_to_load = "amber14/tip3p.xml"
+                    if ff == "tip4p":
+                        ff_to_load = "amber14/tip4pew.xml"
```
If there isn't a way to simplify this I'd suggest wrapping it in a function called `assign_water_forcefield` and sticking it in `utils.py`. I think we should try to keep the code for `parameterize_system` a bit more abstract.

arepstein replied:
Added a lookup dictionary

Commit:
Addressing Comments

Did my best to address all comments

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -360,111 +393,125 @@ def parameterize_system(
+                temp_ff_string = "charmm"
+            elif "amoeba" in ff.lower():
+                temp_ff_string = "amoeba"
+            else:
+                continue
```
what are the implications of continuing here? will a force_field still get assigned?

arepstein replied:
This is to see what forcefield to use in general in order to assign the proper water model. For example, if the user inputs "tip3p," we need to know if the rest of the simulation uses "amber" or "charmm", because they each have their own tip3p: "amber14/tip3p.xml" vs "charmm36/water.xml".
The `continue` is meant to skip assigning a `large_ff_category` if the forcefield specified is a water model. It could be simpler to keep two separate lists, but I was trying to stay consistent with the OpenMM dichotomy

arepstein replied:
Also updated it to not use `continue` anyway

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -360,111 +393,125 @@ def parameterize_system(
+        ff_to_load = None
+        water_assignment = {
+            "amber": {"spce": "amber14/spce.xml", "tip3p": "amber14/tip3p.xml", "tip4p": "amber14/tip4pew.xml"},
+            "charmm": {"spce": "charmm36/spce.xml", "tip3p": "charmm36/water.xml", "tip4p": "charmm36/tip4pew.xml"},
```
I really like the use of a lookup dict

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -360,111 +393,125 @@ def parameterize_system(
-    raise NotImplementedError(
-        f"currently only these force fields are supported: {' '.join(supported_force_fields)}.\n"
-        f"Please select one of the supported force fields."
+            if ff in basic_water_ffs:
+                # Ensure the water model matches the large molecule model
+                if large_ff_category:
+                    if ff in water_assignment[large_ff_category].keys():
+                        ff_to_load = water_assignment[large_ff_category][ff]
+                    else:
+                        warnings.warn(f"Did you mean to use {ff} with the " f"{large_ff_category} force field?")
+                # If there isn't a large molecule forcefield required,
+                # assume amber14
+                else:
+                    ff_to_load = water_assignment["amber"][ff]
+            # TODO: Add lookup to ensure the ff is allowable
+            else:
+                ff_to_load == ff
+            forcefield_omm.loadFile(ff_to_load)
+        # Add small molecules to forcefield
+        for mol, ff_name in small_molecules.items():
+            template = assign_small_molecule_ff(molecules=[mol], forcefield_name=ff_name)
+            forcefield_omm.registerTemplateGenerator(template.generator)
```
I think the `assign_small_molecule_ffs` function is great. Maybe break off a `assign_large_molecule_ffs` and `assign_water_ffs` too?

In general, I think more compartmentalization is better. It would be great if `_parameterize_system` read just invoked the `assign_x_ffs` functions and didn't do any of the dirty work itself. It would be self documenting, where the function names themselves document the purpose of the code.

arepstein replied:
Makes sense! I'll make those changes

orionarcher replied:
Great, looks much cleaner now.

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -360,111 +393,125 @@ def parameterize_system(
+                large_ff_category = temp_ff_string
+            else:
+                if large_ff_category != temp_ff_string:
+                    warnings.warn(f"Did you mean to mix {temp_ff_string} and " f"{large_ff_category} force fields?")
```
see below RE: `assign_large_molecule_ffs`

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -345,6 +345,39 @@ def assign_charges_to_mols(
     return charged_mols
 
 
+def assign_small_molecule_ff(molecules: List[openff.toolkit.topology.Molecule], forcefield_name: str):
```
nice, very clean

orionarcher commented on pymatgen/io/openmm/tests/test_generators.py:
```diff
@@ -73,3 +73,22 @@ def test_get_input_set_w_charges(self):
             "state.xml",
         }
         assert input_set.validate()
+
+    def test_get_input_set_w_charges_and_forcefields(self):
```
I think we should add a more diverse set of tests and test more of the possible forcefields. This would be a great case for [pytest parameterize](https://docs.pytest.org/en/6.2.x/parametrize.html), which would let you test a long list of forcefield inputs with almost complete code re-use.

I also think we should make sure that this is actually doing what we expect. Currently, we are just testing to make sure that it doesn't fail, but there is no confirmation that the Systems are actually parameterized differently. 

We could test parameterization by instantiating a system using `input_set['system.xml'].get_system()` and then using the `System.getForces()` iterator to loop through the forces. 

Even just looping through all the forces and confirming that they are different for different force fields is enough. I'd recommend testing this with just a lone water molecule and with a lone ethanol (or any organic). That'll keep the resulting `Forces` arrays small. Just generating the `Forces` and then using those values as a test would be sufficient, no need to look up in the Sage or GAFF docs what the values are supposed to be.

arepstein replied:
Sounds great, this will take me a bit longer to get to but at the very least will have it done by next week

orioncohen commented:
The changes look great. Thanks for implementing. I have a few more requests in the same vein and some suggestions for other tests to add.

Commit:
Separating functions

Added function to assign_biopolymer_and_water_ff

Commit:
Adding coveralls

orionarcher commented on pymatgen/io/openmm/utils.py:
```diff
@@ -360,111 +444,84 @@ def parameterize_system(
-            if ff_name.lower() in small_ffs:
-                small_molecules[openff_mol] = ff_name.lower()
+        # Add charges to the molecule if provided
+        charged_mols = assign_charges_to_mols(
+            smile_strings=smile_strings,
+            partial_charges=partial_charges,
+            partial_charge_method=partial_charge_method,
+            partial_charge_scaling=partial_charge_scaling,
+        )
```
can we pull this out front and call it once for small molecules and biopolymer_or_water?

orionarcher commented on .coveralls.yml:
```diff
@@ -0,0 +1 @@
+repo_token: KpgnfDK8BDSSTSS4lSrCvHwHnchLgIQ0D
```
Will this link up with a GitHub Hook? I don't quite understand how to get coveralls working, definitely a good idea though.

arepstein replied:
This was an attempt to get it working. Still fairly confused but will try and get it to work. It's an action on materialsproject/pymatgen, and I don't understand why that isn't run here

orionarcher replied:
Ahh, that would have to be set up on my repo. That's my responsibility so for now I'd just remove this.

Later on, (read: post-qual) we'll break out all this code into a separate repo. At that point, we'll set up CI + coveralls.

Commit:
Trying to get linting

Commit:
Unit tests + tip4p removal

Adding unit tests for:
1. smiles with formal charge
2. Ensuring different water models assign different parameters
3. Testing mixing of small molecule ff with water ff

Also removed tip4p support because it requires the addition of an extra particle. Need to figure that out

Commit:
pytest fixing

Fixed smiles string in test_formal_charge
Be careful about smiles!!

Commit:
black formatting

ran black
